### PR TITLE
Pym.js AMP-compatibility

### PR DESCRIPTION
**Pull request by claudiulodro**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR adds basic AMP-compatibility to Pym.js embeds. This is a "good enough" solution until native pym.js AMP compatibility. In general, embeds are designed rectangular, so there will be a little extra padding underneath the amp-iframe, as this renders a square container. We don't have access to the intended height of the iframe from only the src url, and we don't know the exact rectangle ratio so square is a safe choice.

An alternate approach I considered that would always "work correctly" is to detect whether a page is using Pym.js embeds, and if the page is using Pym.js embeds force the page to render in non-AMP. That is also not super ideal, but let me know if you think that's a better solution.

### How to test the changes in this Pull Request:

1. Install and activate Pym.js plugin through the Newspack area of the Plugins screen.
2. Add the following raw content (in details) to a page:

<details>

```
<!-- wp:paragraph -->
<p><a title="Atque." href="http://hills.com/soluta-corrupti-inventore-in-nulla-nisi-fuga">Minima assumenda sed est natus ut.</a> quidem optio neque eum Odio sed sapiente ut recusandae molestiae. Sed laboriosam quas veniam. Error soluta vel Est eum dicta nesciunt quaerat saepe. Quaerat aut et officia Esse deserunt voluptate ea possimus Nihil quia ut doloribus dicta repudiandae. Dolorem soluta perferendis odit. Quia ut consequuntur eum ea harum. Quia expedita voluptates repellat omnis est. Corporis cum adipisci ut. Laboriosam voluptatibus eligendi recusandae ducimus molestias. voluptas adipisci ipsam. Corporis voluptatibus unde facilis repellendus quia neque <a title="Expedita ipsam alias nisi tenetur dolorem." href="http://www.hahn.com/nemo-tenetur-debitis-porro-aut-perferendis-esse-facere">Impedit illo non dignissimos nemo</a></p>
<!-- /wp:paragraph -->

<!-- wp:html -->
<div id="senate" data-pym-src="https://s3.us-west-2.amazonaws.com/projects.oklahomawatch.org/oklahoma-legislature/senate.html"></div>
<!-- /wp:html -->

<!-- wp:paragraph -->
<p>Quasi nihil in assumenda qui amet eum Et ut qui iusto maiores. Rem sint unde quasi. Earum pariatur sunt sed consequatur <a href="http://luettgen.com/enim-molestias-hic-officiis-incidunt-corporis-vel">Expedita</a> omnis harum <a href="http://fay.com/dolor-libero-enim-ducimus-ut.html">dolores. Blanditiis</a> temporibus rerum debitis quia ipsam Quaerat distinctio est voluptatem aut expedita. <a href="http://hane.info/">rerum ut numquam laborum</a> Aut temporibus harum Sed suscipit ut nam. Excepturi aut assumenda accusantium voluptates. Veniam amet est optio harum et Ut voluptatem inventore enim velit itaque. qui delectus deleniti corporis. <a href="http://price.biz/dolor-dolor-hic-vel-ea-unde-architecto-nesciunt.html">molestiae</a> sed non. Et voluptate sunt dicta sit voluptas deserunt in. beatae natus quos ad. Necessitatibus iste molestiae harum <a href="https://anderson.org/eum-ad-minima-error-numquam-alias.html">Eaque et eum et nulla. Est est ab</a> magnam. Eos eaque eligendi ut illum molestias. Omnis totam quisquam temporibus sit sit. Cum consequatur omnis nobis officia.</p>
<!-- /wp:paragraph -->

<!-- wp:pym-shortcode/pym {"src":"https://s3.us-west-2.amazonaws.com/projects.oklahomawatch.org/oklahoma-legislature/senate.html"} -->
<a href="https://s3.us-west-2.amazonaws.com/projects.oklahomawatch.org/oklahoma-legislature/senate.html" class="wp-block-pym-shortcode-pym">https://s3.us-west-2.amazonaws.com/projects.oklahomawatch.org/oklahoma-legislature/senate.html</a>
<!-- /wp:pym-shortcode/pym -->

<!-- wp:paragraph -->
<p>Quasi nihil in assumenda qui amet eum Et ut qui iusto maiores. Rem sint unde quasi. Earum pariatur sunt sed consequatur <a href="http://luettgen.com/enim-molestias-hic-officiis-incidunt-corporis-vel">Expedita</a> omnis harum <a href="http://fay.com/dolor-libero-enim-ducimus-ut.html">dolores. Blanditiis</a> temporibus rerum debitis quia ipsam Quaerat distinctio est voluptatem aut expedita. <a href="http://hane.info/">rerum ut numquam laborum</a> Aut temporibus harum Sed suscipit ut nam. Excepturi aut assumenda accusantium voluptates. Veniam amet est optio harum et Ut voluptatem inventore enim velit itaque. qui delectus deleniti corporis. <a href="http://price.biz/dolor-dolor-hic-vel-ea-unde-architecto-nesciunt.html">molestiae</a> sed non. Et voluptate sunt dicta sit voluptas deserunt in. beatae natus quos ad. Necessitatibus iste molestiae harum <a href="https://anderson.org/eum-ad-minima-error-numquam-alias.html">Eaque et eum et nulla. Est est ab</a> magnam. Eos eaque eligendi ut illum molestias. Omnis totam quisquam temporibus sit sit. Cum consequatur omnis nobis officia.</p>
<!-- /wp:paragraph -->

<!-- wp:shortcode -->
[pym src='https://s3.us-west-2.amazonaws.com/projects.oklahomawatch.org/oklahoma-legislature/senate.html']
<!-- /wp:shortcode -->
```

</details>
3. View on the frontend in AMP and regular mode. You should see an interactive map each place there is a Pym.js embed.

 .

**Note**: If you're getting errors because of double-defined `newspack_is_amp`, check out [this branch of the theme](https://github.com/Automattic/newspack-theme/pull/398).

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by claudiulodro** [comment]
Offered this PR upstream to the pym.js plugin: https://github.com/INN/pym-shortcode/issues/59#issuecomment-531488484

**Comment by claudiulodro** [comment]
@westonruter if you have a chance, I'd appreciate any thoughts on this approach. It's pretty straightforward. I'd also be curious if you know a good solution to making amp-iframe height equal to the height of the inner iframe's content; something like `layout='fit'` would be cool (good feature idea for the next version of AMP! :D).

File: includes/class-amp-enhancements.php [hunk 1]
```diff
@@ -37,5 +39,68 @@ public static function documentcloud_iframe_full_height( $output, $tag ) {
...
+	 *
+	 * @todo   This is a "good enough" solution until native pym.js AMP compatibility. In general, embeds are designed rectangular,
+	 * so there will be a little extra padding underneath the amp-iframe, as this renders a square container. We don't have access to
+	 * the intended height of the iframe from only the src url.
+	 * @see    https://github.com/ampproject/amphtml/issues/22714
+	 * @param  string $src iframe src.
+	 * @return string AMP-iframe HTML.
+	 */
+	protected static function get_pym_ampiframe( $src ) {
+		return "<amp-iframe src='" . esc_url( $src ) . "' layout='responsive' width='1200' height='1200' sandbox='allow-scripts allow-same-origin' frameborder='0'></amp-iframe>";
```
Comment (westonruter) [inline includes/class-amp-enhancements.php]: In order to preempt Pym.js support, and to support pages that actually send the AMP resizing messages, this `amp-iframe` should include the `resizable` attribute, as well as an `overflow` child element: https://amp.dev/documentation/components/amp-iframe/?format=websites#iframe-resizing

Additionally, to protect situations where the iframe appears in the initial viewport, you should have a `placeholder` child as well (without which the iframe is blocked from rendering, AFAIK): https://amp.dev/documentation/components/amp-iframe/?format=websites#iframe-with-placeholder

**Comment by westonruter** [comment]
> I'd also be curious if you know a good solution to making amp-iframe height equal to the height of the inner iframe's content; something like `layout='fit'` would be cool (good feature idea for the next version of AMP! :D).

Yes, that is supported… but it depends on the document in the iframe window sending this message:

```js
window.parent.postMessage({
  sentinel: 'amp',
  type: 'embed-size',
  height: document.body.scrollHeight
}, '*');
```

Without the embedded window sending such a message, there is no way for the host page to “see” into the window to determine its dimensions. This is why Pym.js exists, actually. 

For more, see the documentation on [`amp-iframe` resizing](https://amp.dev/documentation/components/amp-iframe/?format=websites#iframe-resizing).

**Comment by westonruter** [comment]
It may be possible for you to inject that logic yourself if you have the interactives on the local filesystem:

https://gist.github.com/westonruter/73412f3e35cdeaade808f3b8da9d9586#file-pymjs-amp-injected-js
https://gist.github.com/westonruter/73412f3e35cdeaade808f3b8da9d9586#file-pymjs-testing-php-L96-L119

If the file is located on a remote server requiring HTTP request to fetch, then this would not be a good approach, because the interactive would need to be cached in a transient. When embedding an interactive for things like election results, such added latency is not ideal 😄

File: includes/class-amp-enhancements.php (hunk 1 above)
Comment (claudiulodro) [reply includes/class-amp-enhancements.php]: https://github.com/Automattic/newspack-plugin/pull/276/commits/9cefe4f60edc1b04ef196ab90d2774c26a132274

**Comment by claudiulodro** [comment]
Thanks for the feedback @westonruter! I've implemented the changes you requested and verified the resizing works well on pages that include your above JS snippet.

@jeffersonrabb ping me for instructions if you want to test the resizing.

**Comment by claudiulodro** [comment]
Don't merge this; I'm going to move the functionality to the Pym.js plugin. https://github.com/INN/pym-shortcode/issues/59#issuecomment-531835646

**Comment by claudiulodro** [comment]
https://github.com/INN/pym-shortcode/pull/62

**Comment by benlk** [comment]
We've merged https://github.com/INN/pym-shortcode/pull/62 and are proceeding to get that released to wp.org.

**Comment by benlk** [comment]
Released: https://github.com/INN/pym-shortcode/releases/tag/v1.3.2.3

**Comment by claudiulodro** [comment]
Thanks for getting that AMP-compatibility work released @benlk!

I've removed the handling from here since it's merged upstream, and this PR now just adds Pym.js Embeds to the list of supported plugins.

File: includes/class-plugin-manager.php [hunk 2]
```diff
@@ -296,6 +296,14 @@ public static function get_managed_plugins() {
 				'PluginURI'   => 'https://github.com/xwp/pwa-wp',
 				'Download'    => 'wporg',
 			],
+			'pym-shortcode'                 => [
+				'Name'        => 'Pym.js Embeds',
+				'Description' => 'A WordPress solution to embed iframes that are responsive horizontally and vertically using the NPR Visuals Team\'s Pym.js.',
+				'Author'      => 'INN Labs',
+				'AuthorURI'   => 'http://labs.inn.org/',
```
Comment (benlk) [inline includes/class-plugin-manager.php]: This should be https rather than http, but it's not a big problem.

Comment (claudiulodro) [reply includes/class-plugin-manager.php]: No worries. Updated in https://github.com/Automattic/newspack-plugin/pull/480 for next week's release. 

[It's worth noting that this info is taken directly from the Pym.js Embeds plugin info, and you have http there.](https://github.com/INN/pym-shortcode/blob/master/pym-shortcode.php#L8) :)

Comment (benlk) [reply includes/class-plugin-manager.php]: Oh weird; I thought we had updated that. Thanks for letting me know!

**Pull request merged**